boot/bootstate20.go

```diff
@@ -0,0 +1,339 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/snap"
+)
+
+type bootState20Base struct{}
+
+func (s20 *bootState20Base) revisions() (theSnap, theTrySnap snap.PlaceInfo, trying bool, err error) {
+	// TODO:UC20: implement support for trying with base_status and try_base
+	modeenv, err := ReadModeenv(dirs.GlobalRootDir)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot get snap revision: unable to read modeenv: %v", err)
+	}
+
+	if modeenv.Base == "" {
+		return nil, nil, false, fmt.Errorf("cannot get snap revision: modeenv base boot variable is empty")
+	}
+
+	sn, err := snap.ParsePlaceInfoFromSnapFileName(modeenv.Base)
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot get snap revision: modeenv base boot variable is invalid: %v", err)
```

Comment:
Wonder if it'd make sense to update the common code in `boot.go` and add the `cannot get snap revision` prefix to errors at that level. FWIW, just an idea for a follow up material.